### ISSUE TITLE
Obscure error on TextResponse with binary content-type

### ISSUE THREAD

**Issue by lopuhin** [scrapy-plugins/scrapy-splash#107]
If response is binary according to it's content-type, but can be decoded using some encoding, it becomes ``TextResponse`` in scrapy. But [_change_response_class](https://github.com/scrapy-plugins/scrapy-splash/blob/0e3284abdcc2b4e1b228754f3c8ec76cceac4ec0/scrapy_splash/middleware.py#L394) in the splash middleware looks at the content type, so if it's binary, but response was a ``TextResponse``, the following error occurs:
```
In [1]: response.body
Out[1]: b'\xb8\xfe\xff'

In [2]: response
Out[2]: <200 http://127.0.0.1:8050/execute>

In [3]: type(response)
Out[3]: scrapy.http.response.text.TextResponse

In [4]: response.headers
Out[4]:
{b'Content-Type': b'application/pdf',
 b'Date': b'Thu, 09 Mar 2017 10:02:21 GMT',
 b'Server': b'TwistedWeb/16.1.1'}

In [5]: responsetypes.from_args(headers=response.headers)
Out[5]: scrapy_splash.response.SplashResponse

In [6]: respcls = responsetypes.from_args(headers=response.headers)

In [7]: response.replace(cls=respcls, request=request)
---------------------------------------------------------------------------
_DefGen_Return                            Traceback (most recent call last)
/Users/kostia/shub/memex/undercrawler/venv/lib/python3.5/site-packages/twisted/internet/defer.py in _inlineCallbacks(result, g, deferred)
   1298             else:
-> 1299                 result = g.send(result)
   1300         except StopIteration as e:

/Users/kostia/shub/scrapy/scrapy/core/downloader/middleware.py in process_request(request)
     42                     defer.returnValue(response)
---> 43             defer.returnValue((yield download_func(request=request,spider=spider)))
     44

/Users/kostia/shub/memex/undercrawler/venv/lib/python3.5/site-packages/twisted/internet/defer.py in returnValue(val)
   1275     """
-> 1276     raise _DefGen_Return(val)
   1277

_DefGen_Return: <200 http://127.0.0.1:8050/execute>

During handling of the above exception, another exception occurred:

TypeError                                 Traceback (most recent call last)
/Users/kostia/shub/scrapy-splash/scrapy_splash/middleware.py in <module>()
----> 1 response.replace(cls=respcls, request=request)

/Users/kostia/shub/scrapy/scrapy/http/response/text.py in replace(self, *args, **kwargs)
     52     def replace(self, *args, **kwargs):
     53         kwargs.setdefault('encoding', self.encoding)
---> 54         return Response.replace(self, *args, **kwargs)
     55
     56     @property

/Users/kostia/shub/scrapy/scrapy/http/response/__init__.py in replace(self, *args, **kwargs)
     79             kwargs.setdefault(x, getattr(self, x))
     80         cls = kwargs.pop('cls', self.__class__)
---> 81         return cls(*args, **kwargs)
     82
     83     def urljoin(self, url):

/Users/kostia/shub/scrapy-splash/scrapy_splash/response.py in __init__(self, url, *args, **kwargs)
     31                 self.real_url = url
     32                 url = _url
---> 33         super(_SplashResponseMixin, self).__init__(url, *args, **kwargs)
     34
     35     def replace(self, *args, **kwargs):

TypeError: __init__() got an unexpected keyword argument 'encoding'

In [8]: response.encoding
Out[8]: 'cp1252'
```

I'm not sure what would be the best fix in this case: making sure ``SplashTextResponse`` is used if original response is ``TextResponse``?

**Comment by kmike** [comment]
I'm also not sure about the fix, but your proposed solution sounds ok.

**Comment by eusid** [comment]
I was waiting on this pull request in hopes it would fix my issue. Please see my stackoverflow post. http://stackoverflow.com/questions/43090352/proxy-servers-with-scrapy-splash  At least the error I'm getting is different :) You'd be my savior if you could tell me what I'm doing wrong.

**Comment by vionemc** [comment]
One known issue from http://stackoverflow.com/questions/43090352/proxy-servers-with-scrapy-splash 
I've checked, and yeah, the error is gone when I turn off Crawlera. It's so sad it doesn't work well with Crawlera.
